src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,289 @@
+
+            // work queue for refreshing LSP data
+            _documentSymbolQueue = new AsyncBatchingResultQueue<DocumentSymbolDataModel?>(
+                DelayTimeSpan.Short,
+                GetDocumentSymbolAsync,
+                asyncListener,
+                CancellationToken);
+
+            // work queue for updating UI state
+            _updateViewModelStateQueue = new AsyncBatchingWorkQueue<ViewModelStateDataChange>(
+                DelayTimeSpan.Short,
+                UpdateViewModelStateAsync,
+                asyncListener,
+                CancellationToken);
+
+            _taggerEventSource.Changed += OnEventSourceChanged;
+            _taggerEventSource.Connect();
+
+            // queue initial model update
+            _documentSymbolQueue.AddWork(cancelExistingWork: true);
+        }
+
+        private SortOption _sortOption = SortOption.Location;
+        public SortOption SortOption
+        {
+            get => _sortOption;
+            set
+            {
+                _threadingContext.ThrowIfNotOnUIThread();
+                SetProperty(ref _sortOption, value);
+            }
+        }
+
+        private string? _searchText;
+        public string? SearchText
+        {
+            get => _searchText;
+            set
+            {
+                _threadingContext.ThrowIfNotOnUIThread();
+                _searchText = value;
+                EnqueueFilter(_searchText);
+            }
+        }
+
+        private ImmutableArray<DocumentSymbolDataViewModel> _documentSymbolViewModelItems = ImmutableArray<DocumentSymbolDataViewModel>.Empty;
+        public ImmutableArray<DocumentSymbolDataViewModel> DocumentSymbolViewModelItems
+        {
+            get => _documentSymbolViewModelItems;
+            set
+            {
+                _threadingContext.ThrowIfNotOnBackgroundThread();
+
+                // Unselect any currently selected items or WPF will believe it needs to select the root node.
+                DocumentOutlineHelper.UnselectAll(_documentSymbolViewModelItems);
+                SetProperty(ref _documentSymbolViewModelItems, value);
+            }
+        }
+
+        private void OnEventSourceChanged(object sender, TaggerEventArgs e)
+            => _documentSymbolQueue.AddWork(cancelExistingWork: true);
+
+        public event PropertyChangedEventHandler? PropertyChanged;
+
+        private void NotifyPropertyChanged([CallerMemberName] string? propertyName = null)
+            => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
+
+        private void SetProperty<T>(ref T field, T value, [CallerMemberName] string propertyName = "")
+        {
+            if (EqualityComparer<T>.Default.Equals(field, value))
+            {
+                return;
+            }
+
+            field = value;
+            NotifyPropertyChanged(propertyName);
+        }
+
+        public void Dispose()
```

Comment:
nit: preference it to keep Dispose paired with constructor (i.e. it should appear after it).